include/zephyr/pm/pm.h

```diff
@@ -53,18 +53,136 @@ extern "C" {
  */
 struct pm_notifier {
 	sys_snode_t _node;
+
 	/**
-	 * Application defined function for doing any target specific operations
+	 * Value indicating whether the callback should trigger on entry or
+	 * exit of a power state.
+	 */
+	enum pm_state_direction direction;
+
+	/**
+	 * Application or driver defined function for doing any target specific operations
 	 * for power state entry.
 	 */
-	void (*state_entry)(enum pm_state state);
+	void (*callback)(void *arg1, void *arg2, void *arg3);
+
 	/**
-	 * Application defined function for doing any target specific operations
-	 * for power state exit.
+	 * Optional callback arguments
 	 */
-	void (*state_exit)(enum pm_state state);
+	void *arg1;
+	void *arg2;
+	void *arg3;
 };
 
+/**
+ * @brief Helper macro for building a pm_notifier list name.
+ *
+ * @param node_id CPU node identifier
+ * @param idx Index of the list property containing power-state phandles
+ */
+#define Z_PM_NOTIFIERS_NAME(node_id, idx)				\
+	UTIL_CAT(pm_notifiers_cpu_,					\
+		 UTIL_CAT(DT_NODE_CHILD_IDX(node_id),			\
+			  UTIL_CAT(_state_, idx)))
+
+/**
+ * @brief Helper macro for initializing a pm_notifiers list for a specific CPU's
+ *	  cpu-power-states power-state phandle.
+ *
+ * @param node_id CPU node identifier
+ * @param prop Property containing one or more power-state phandles
+ * @param idx Index of the list property containing power-state phandles
+ */
+#define Z_PM_NOTIFIERS_CREATE_STATES(node_id, prop, idx)		\
+	static sys_slist_t Z_PM_NOTIFIERS_NAME(node_id, idx)		\
+		= SYS_SLIST_STATIC_INIT(&				\
+			Z_PM_NOTIFIERS_NAME(node_id, idx))
+
+/**
+ * @brief Generate individual, per power-state pm_notifiers lists for the
+ *	  passed CPU id.
+ *
```

Comment:
It shouldn't be per CPU imho, if you have a list per CPU you will probably need to have N notifiers copy (one per cpu). Unless you have specific needs per cpu what you really want to know is when a state is used and in this case  you would need to have a copy per cpu.

Reply:
I'll do some more research into Zephyr's SMP implementation, but I think I understand your point.

What do you think of a CPU-independent array of lists per `power-states` node? That would also significantly simplify the macros added here, and should avoid SMP issues.

Reply:
Yep, that is my point. Just keep lists per states (or one single list if this is not really a bottleneck)

Reply:
Done.